mpc85xx: add support for Aerohive AP370

dmascord opened the pull request:
mpc85xx: add support for Aerohive AP370

    This patch adds support for the Aerohive AP370.
    It includes support for sysupgrade installation of dtb, ramdisk and kernel (nand tar devices only)

    Specifications:
     * SoC: QorIQ P1020 e500v2
     * RAM: 512 MB
     * Flash: 1024 MB NAND flash
     * WiFi: Atheros QCA9890 (5 GHz) and AR9590 (2.4 GHz)
     * Ethernet: 2 x 1000Base-T (PoE)
     * LED: One combined activity LED (orange/white/off)
     * UART: 9600,8,N,1 RJ45 console

    Installation:
     * Installation via console serial tftp over ethernet
       + First download the patched uboot from https://drive.google.com/drive/folders/1Hq4QTwh5FlzVbmrsL4JoKNmdnB4lsJ__?usp=sharing
       + Copy patched uboot into tftp folder
       + Create Serial to RJ45 cable, or purchase premade rollover cable, as per user guide at https://fccid.io/png.php?id=1000302&page=6
       + Connect console cable to USB-RS232 or native serial port on old PC, and connect to AP370 console RJ45 port
       + Use miniterm/putty etc to connect 9600,8,N,1 (miniterm.py /dev/ttyS0 --raw --eol CR)
       + Power AP370 with PoE or 12V DC barrel jack
       + Hit space to interupt the boot process
       + Enter uboot password AhNf?d@ta06 or administrator
       + Enter the following commands to boot from tftp image, changing 192.168.1.1 to the tftp server:
         dhcp;
         setenv serverip 192.168.1.1;
         tftpboot 0x1000000 openwrt-mpc85xx-p1020-aerohive_hiveap-370-initramfs-kernel.bin;
         tftpboot 0x6000000 openwrt-mpc85xx-p1020-aerohive_hiveap-370-squashfs-fdt.bin;

         tftpboot 0x8000000 mtd1_Uboot.patched
         nand erase 0x20000 0x200000
         nand write 0x8000000 0x20000 0x200000

         tftpboot 0x8000000 mtd2_Uboot-backup.patched
         nand erase 0x220000 0x200000
         nand write 0x8000000 0x220000 0x200000

         setenv owrt_boot_a 'setenv bootargs root=/dev/ram console=ttyS0,9600 ramdisk_size=700000 cache-sram-size=0x10000\; \
    nand read 0x01005000 0x00520000 0x4000000\; nand read 0x08005000 0x04520000 0x00080000\; bootm 0x01005000 - 0x08005000\;'
         saveenv

         bootm 0x1000000 - 0x6000000;
       + Once booted, scp across the sysupgrade image and upgrade:
         sysupgrade -v /tmp/openwrt-mpc85xx-p1020-aerohive_hiveap-370-squashfs-sysupgrade.bin
       + The initial format of the ubi filesystem will take a long time, so please be patient
       + If it boots into the backup partition (on factory), you will need to hit the reset button to switch it from boot image 1 or 0

    Signed-off-by: Damien Mascord <tusker@tusker.org>

blocktrron commented on package/base-files/files/lib/upgrade/nand.sh:
```diff
@@ -7,6 +7,12 @@
+CI_DTBPART="${CI_DTBPART:-dtb}"
+
+# 'ramdisk' partition on NAND contains the ramdisk
+CI_RAMDISKPART="${CI_RAMDISKPART:-ramdisk}"
```
The Ramdisk is fake - do you actually need to update this partition when performing sysupgrade?

dmascord replied:
When I tried to leave the ramdisk as per factory image, it would fail to boot.
When I wiped the partition, it also failed to boot.
When I put a fake ramdisk, it allowed it to boot.
When I tried to change the uboot boot parameters so that it didn't need the ramdisk, it ignored the environment settings and still attempted to use the ramdisk.

adrianschmutzler replied:
I don't know so much about that script, but again I wonder whether this has to be done for entire OpenWrt instead of just applying it on a more local basis.

dmascord replied:
The script will only execute the portions of the nand if a) the image is present b) the mtd is labelled, but I understand your concern.
I can test the changes on other nand devices that I have, if you wish ?

Hurricos replied:
Changes to `package/base-files/files/lib/upgrade/nand.sh` have been full dropped.

blocktrron commented on target/linux/mpc85xx/base-files/etc/board.d/02_network:
```diff
@@ -10,7 +10,7 @@ board_config_update
 
 case "$board" in
-aerohive,hiveap-330)
+aerohive,hiveap-330,hiveap-370)
```
```
aerohive,hiveap-330|\
aerohive,hiveap-370)
```

blocktrron commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -15,6 +15,27 @@ define Device/aerohive_hiveap-330
+  DEVICE_MODEL := HiveAP-370
+  DEVICE_PACKAGES := kmod-tpm-i2c-atmel
+  BLOCKSIZE := 128k
+  # does not support FIT
+  # KERNEL := kernel-bin | gzip | fit gzip $(KDIR)/image-$$(DEVICE_DTS).dtb
+  # does not support appended dtb with lzma
+  # KERNEL := kernel-bin | append-dtb | lzma
+  # does not support combined dtb (Verifying Checksum ... Bad Data CRC)
+  # KERNEL := kernel-bin | append-dtb | uImage gzip
```
You can drop all of these and replace it by a one liner "FIT and LZMA are not supported".

After all, `append-dtb` relies on legacy DTB booting being enabled for the kernel. I'm not sure if PPC even supports this.

dmascord replied:
Probably can remove these comments, as they only serve for informational purposes.

adrianschmutzler commented on include/image-commands.mk:
```diff
@@ -329,6 +330,8 @@ define Build/sysupgrade-tar
 		--kernel $(call param_get_default,kernel,$(1),$(IMAGE_KERNEL)) \
 		--rootfs $(call param_get_default,rootfs,$(1),$(IMAGE_ROOTFS)) \
+		--dtb $(call param_get_default,dtb,$(1),$(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb) \
+		--ramdisk $(call param_get_default,ramdisk,$(1),$(IMAGE_KERNEL).fakehdr) \
```
This introduces changes to all devices in OpenWrt using that line. Please don't do that (unless you think that _all_ devices require it and you have tested a fair share of them. In that case, it should be a separate commit BTW.)
If you need this for a specific device, please create a specific recipe for it (in the relevant target's image/Makefile).
Same goes for the addition above in append-uImage-fakehdr

dmascord replied:
OK, let me try to rework the patch to make it device specific.  If more than one device ends up needing it, we can generalize it.

Hurricos replied:
Changes to `include/image-commands.mk` have been fully dropped.

adrianschmutzler commented on package/base-files/files/lib/upgrade/nand.sh:
```diff
@@ -258,9 +268,20 @@ nand_upgrade_tar() {
+	[ "$ramdisk_length" != 0 -a -n "$ramdisk_mtd" ] && {
+		tar xf $tar_file ${board_dir}/ramdisk -O | mtd -e $CI_RAMDISKPART write - $CI_RAMDISKPART
+		sleep 1
+	}
```
See above.

dmascord replied:
OK

adrianschmutzler commented on target/linux/mpc85xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -14,7 +14,8 @@ platform_do_upgrade() {
 	ocedo,panda|\
-	sophos,red-15w-rev1)
+	sophos,red-15w-rev1|\
+	aerohive,hiveap-370)
```
Please care about alphabetic sorting.

dmascord replied:
OK

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,294 @@
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ */
```
Please use an SPDX license identifier instead.

dmascord replied:
OK

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,294 @@
+			label = "ap370:white:status";
+			gpios = <&gpio0 11 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+			linux,default-trigger = "default-on";
```
Does that trigger exist?

dmascord replied:
From what I can see, it does exist:

root@OpenWrt:~# cat /sys/class/leds/ap370\:white\:status/trigger
[none] timer default-on netdev phy0rx phy0tx phy0assoc phy0radio phy0tpt phy1rx phy1tx phy1assoc phy1radio phy1tpt

adrianschmutzler replied:
Interesting. I'm used to just using the default-state property here.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,294 @@
+	        };
+	};
+
+	board_soc: soc: soc@ffe00000 {
```
Do these dual labels work?

dmascord replied:
the dtc compiler doesn't complain about it, but it is probably better to clean it up and only have one reference.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,294 @@
+					rtc@68 {
+							compatible = "dallas,ds1339";
+							reg = <0x68>;
+					};
```
broken indent.

dmascord replied:
OK

Hurricos replied:
Fixed.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,294 @@
+			interrupt-parent = <0x2>;
+			gpio-controller;
+			available = <0x3 0x5 0x8 0xa 0xb 0xc 0xd 0xf>;
+	    };
```
broken indent.

dmascord replied:
OK

adrianschmutzler commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -15,6 +15,27 @@ define Device/aerohive_hiveap-330
 endef
 TARGET_DEVICES += aerohive_hiveap-330
 
+define Device/aerohive_hiveap-370
```
One should consider creating a shared Device/aerohive_hiveap definition here.

dmascord replied:
Is there a good example I can follow ?

adrianschmutzler replied:
Something like that, only you can keep it all together and don't have to spread it across files:
https://github.com/openwrt/openwrt/commit/3b89849b57a88d0919b6442509131f045af580a8

Hurricos replied:
OK, @adschm; your request is now done:

```diff
@@ -1,13 +1,19 @@                                                                                                                                                                               
-define Device/aerohive_hiveap-330                                                                                                                                                              
+define Device/aerohive_hiveap                                                                                                                                                                  
   DEVICE_VENDOR := Aerohive                                                                                                                                                                    
-  DEVICE_MODEL := HiveAP-330                                                                                                                                                                   
   DEVICE_PACKAGES := kmod-tpm-i2c-atmel                                                                                                                                                        
   BLOCKSIZE := 128k                                                                                                                                                                            
   KERNEL := kernel-bin | gzip | uImage gzip                                                                                                                                                    
-  KERNEL_SIZE := 8m                                                                                                                                                                            
   KERNEL_INITRAMFS := copy-file $(KDIR)/vmlinux-initramfs | uImage none                                                                                                                        
-  IMAGES := fdt.bin sysupgrade.bin                                                                                                                                                             
   IMAGE/fdt.bin := append-dtb                                                                                                                                                                  
+  IMAGES := fdt.bin sysupgrade.bin kernel.bin rootfs.bin                                                                                                                                       
+  IMAGE/kernel.bin := append-kernel                                                                                                                                                            
+  IMAGE/rootfs.bin := append-rootfs | pad-rootfs                                                                                                                                               
+endef                                                                                                                                                                                          
+                                                                                                                                                                                               
+define Device/aerohive_hiveap-330                                                                                                                                                              
+  $(Device/aerohive_hiveap)                                                                                                                                                                    
+  DEVICE_MODEL := HiveAP-330                                                                                                                                                                   
+  KERNEL_SIZE := 8m            
```

adrianschmutzler commented on target/linux/mpc85xx/p1020/config-default:
```diff
@@ -54,3 +53,5 @@ CONFIG_UBIFS_FS_ZLIB=y
 CONFIG_WS_AP3710I=y
 CONFIG_XPS=y
+CONFIG_HIVEAP_330=y
+CONFIG_HIVEAP_370=y
```
Sorting.

dmascord replied:
A question here actually, I don't really want to define CONFIG_HIVEAP_330 when selecting the device configuration, but if I don't define both 330 and 370 here, it asks every build if I want to include either 330 or 370

adrianschmutzler replied:
HIVEAP_330 is already defined in that file, you just move it down.

dmascord commented:
Made the dtb and ramdisk components optional, so that only devices that want them will be included.
Made a common Device definition for aerohive_hiveap between the AP330 and AP370, since they are very similar.

dmascord commented:
Updated the sysupgrade nand.sh to use nandwrite (from nand-utils), which fixes the mtd write issue where the dtb failed to write and produced only ff ff ff ff written.  With nandwrite, it is no longer necessary to sleep 1, which wasn't reliable anyway.  This makes it seem like there is an issue in the mtd layer.

dmascord commented:
Just as an update here, investigation into the nand write issue has revealed that the offsets between uboot and linux do not match (for some mtd partitions), which leads me to believe there is an issue with the nand driver.

This relates very closely to what is seen on https://bugs.openwrt.org/index.php?do=details&task_id=1926, so a similar fix may need to be done for fsl,elbc-fcm-nand.

From what I can see, the data is being written to the NAND properly by Linux, but is misaligned by 0x60000, therefore uboot fails to read the dtc or ramdisk during startup.

adschm commented on include/image-commands.mk:
```diff
@@ -124,6 +124,7 @@ define Build/append-uImage-fakehdr
 		-s \
 		$@.fakehdr
 	cat $@.fakehdr >> $@
+	cat $@.fakehdr >> $(IMAGE_KERNEL).fakehdr
```
Err, this changes behavior for _all_ devices using that recipe...
This should be made in a separate commit, and somebody should check the effect on the relevant devices ...

Hurricos replied:
All set; all changes to `include/image-commands.mk` are now dropped.

adschm commented on include/image-commands.mk:
```diff
@@ -332,6 +333,18 @@ define Build/sysupgrade-tar
+		--dtb $(call param_get_default,dtb,$(1),$(dtb_file)) \
+		--ramdisk $(call param_get_default,ramdisk,$(1),$(ramdisk_file)) \
+		$@
+endef
```
Do you plan to use this for other devices? If not, you should keep it in target/linux/mpc85xx/image/Makefile.

Hurricos replied:
Dropped!

adschm commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -1,20 +1,34 @@
+  DEVICE_MODEL := HiveAP-370
+  DEVICE_PACKAGES += kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct kmod-ath9k nand-utils
+  IMAGES := fdt.bin ramdisk.bin sysupgrade.bin
+  IMAGE/ramdisk.bin := append-uImage-fakehdr ramdisk | pad-to 256k | check-size 512k |
```
Pipe at the end?

Hurricos replied:
Cleaned.

adschm commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -1,20 +1,34 @@
+define Device/aerohive_hiveap-370
+  $(Device/aerohive_hiveap)
+  DEVICE_MODEL := HiveAP-370
+  DEVICE_PACKAGES += kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct kmod-ath9k nand-utils
```
smallbuffers on a device with 512 MB RAM?

Hurricos replied:
Ew, no need. Cleaned, now.

adschm commented on target/linux/mpc85xx/p1020/config-default:
```diff
@@ -54,3 +55,4 @@ CONFIG_UBIFS_FS_ZLIB=y
 CONFIG_UBIFS_FS_ZSTD=y
 CONFIG_WS_AP3710I=y
 CONFIG_XPS=y
+
```
Remove that empty line again.

dmascord commented:
The same uboot vs linux offset issue is faced on other NAND devices, which looks like due to an ECC strength issue rather than BBT - https://www.yoctoproject.org/pipermail/meta-freescale/2017-April/020344.html 

I am not familiar enough with fsl_elbc_nand.c to know how the ECC or BBT offsets are causing issues with the MTD geometry, so any help would be appreciated.

adschm commented:
This has been open for a long time already, and at least two major concerns have not been addressed during that time. In addition, the chances of this being picked up eventually are small even if those are fixed. So, I think it's better to close this now ...

dmascord commented:
It looks like the flash on the test device I was using was bad, and two other AP370 that I tested didn't have this problem at all, so I am going to clean up this PR and make it acceptable again, should I reopen this ?

dmascord commented:
@adschm I have reverted the commit as per the wiki entry, could you please reopen ?

dmascord commented:
@adschm - I've pushed the changes again, based on the previous review comments, which I believe covers all issues.  Sysupgrade is working fine on the two AP370s that I have at the moment.

Hurricos commented:
@adschm: Due to the combination of 'forced' `bootcmd` and a limited gunzip space, would it be palatable to require users download a patched version of U-boot and replace it, or would it be preferable to provide code to manually patch out the U-boot script string which is run to replace the `bootcmd`?

Hurricos reviewed (commented):
Added some commentary. I will act upon it and test a build sometime today.

Hurricos commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -1,20 +1,35 @@
+define Device/aerohive_hiveap-370
+  $(Device/aerohive_hiveap)
+  DEVICE_MODEL := HiveAP-370
+  DEVICE_PACKAGES += kmod-ath10k-ct ath10k-firmware-qca988x-ct kmod-ath9k nand-utils
```
This also needs the `block-mount` package in order to mount the `rootfs_data` volume.

Hurricos replied:
`block-mount` has now been added; `rootfs_data` now automounts.

Hurricos commented:
Hi @adschm: I have significantly refactored on this pull request; I will move `dmascord:add_aerohive_ap370` out of the way and force-push to it with this refactoring.

Hurricos committed:
mpc85xx: Add Aerohive HiveAP-370 Access Point

This patch adds support for the Aerohive HiveAP 370. It is very similar to
the HiveAP-330.

Specifications:
 * SoC: QorIQ P1020 e500v2 @ 800MHz
 * RAM: 512 MB
 * Flash: 1024 MB NAND flash
 * WiFi: Atheros QCA9890 (5 GHz) and AR9590 (2.4 GHz)
 * Ethernet: 2 x 1000Base-T (PoE)
 * LED: One combined activity LED (orange/white/off)
 * UART: 9600,8,N,1 RJ45 console

Installation:
 * Installation via console serial tftp over ethernet
   + Connect to the RJ45-based serial console port (see https://openwrt.org/docs/techref/hardware/port.serial)
   + Use miniterm/putty etc to connect 9600,8,N,1 (miniterm.py /dev/ttyS0 --raw --eol CR)
   + Power AP370 with PoE or 12V DC barrel jack
   + Hit space to interupt the boot process
   + Enter uboot password AhNf?d@ta06 or administrator
   + Enter the following commands to boot from tftp image:

     dhcp;
     setenv serverip <TFTP server IPv4 address>;
     tftpboot 0x1000000 openwrt-mpc85xx-p1020-aerohive_hiveap-370-initramfs-kernel.bin;
     tftpboot 0x6000000 openwrt-mpc85xx-p1020-aerohive_hiveap-370-squashfs-fdt.bin;

     setenv owrt_boot_a 'nand read 0x6000000 0x00520000 0x40000\; nand read 0x1000000 0x560000 0x02000000\; bootm 0x1000000 - 0x6000000\;'
     saveenv;

     bootm 0x1000000 - 0x6000000;

   + Once booted, patch your u-boot installation so that it runs
     `owrt_boot_a` on boot by doing:

     . /lib/upgrade/patch-aerohive-uboot.sh;
	 patch_aerohive_hiveap_370_uboot;

   + Finally, wget or scp over the sysupgrade file under /tmp, and
     sysupgrade with:

     sysupgrade -v /tmp/openwrt-mpc85xx-p1020-aerohive_hiveap-370-squashfs-sysupgrade.bin

Notes:

* The initial format of the ubi filesystem will take a long time, so
  please be patient.

* A/B partitioning functionality has been dropped: both copies of
  U-boot run `owrt_boot_a` to boot.

* There is no 'official' way to possible to revert to the original
  firmware, but this should be of no concern, as the device is EoL on
  the original FW:
  https://aerohive-www-cdn.aerohive.com/wp-content/uploads/Aerohive_EOS-EOL_AP370-AP390.pdf

* The kernel partition contains a shimmed dtb to allow for the use of
  sysupgrade-tar.

* The MAC address is pulled from NAND flash using nvmem

  We actually need to enclose the whole section of partitions in a
  `partitions { ... }` because we need to assign at least the hwinfo
  partition a `compatible = "fixed-partitions";`; otherwise it won't
  be registered when bringing up MTD, as per these links:

  + https://forum.openwrt.org/t/tp-link-c2600-missing-default-mac-mtd-partition-in-snapshot/103945/6

  + Commit 67efc978cce2

  + Commit 11d425f1863b

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

Hurricos commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,295 @@
+	};
+
+	chosen {
+		bootargs-override = "console=ttyS0,9600";
```
Hard-setting the `ubi.mtd` is not necessary to automatically mount: only the `block-mount` package was needed.

Hurricos replied:
I have dropped the hard-setting of ubi.mtd.

Hurricos commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/hiveap-370.dts:
```diff
@@ -0,0 +1,295 @@
+							label = "Hardware Info";
+					};
+					
+					partition@520000 {
```
In order to allow the `rootfs` and `rootfs_data` to go into UBI, I found a better solution was to make use of the existing `nand.sh`'s behavior (by using a `sysupgrade-tar`):

`append-dtb | pad-to 256k | append-kernel | sysupgrade-tar kernel=$$$$@ | append-metadata`

Hurricos replied:
I ended up using this solution to shim the dtb before the beginning of the kernel.

This is needed because the bootloader does not accept FIT images, and cannot find the dtb if it is appended to the kernel: in this case, we instruct u-boot to load the dtb and kernel from fixed addresses, and boot from them directly.

This solution lets us also load rootfs and rootfs_data from, and sysupgrade to, UBI partitions, despite this board's u-boot's lack of UBI support.

dmascord commented:
@adschm - merged in Hurricos's changes into this PR.

@Hurricos thanks for the effort on cleaning up this PR!

Hurricos commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -1,20 +1,34 @@
+define Device/aerohive_hiveap-370
+  $(Device/aerohive_hiveap)
+  DEVICE_MODEL := HiveAP-370
+  DEVICE_PACKAGES += kmod-tpm-i2c-atmel kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct kmod-ath9k nand-utils
```
Done.

Hurricos commented:
I realized I now have the power to force-push to `dmascord/openwrt`, so I've done so.

Note that as of yesterday, this PR has two commits: one adding the HiveAP 370 itself, and another to fix a [bug in SMP initialization which was introduced in linux-stable 5.10.80](https://lists.ozlabs.org/pipermail/linuxppc-dev/2021-November/237026.html). The [patch is awaiting a merge upstream](https://lists.ozlabs.org/pipermail/linuxppc-dev/2021-November/237134.html).

chunkeey reviewed (commented):
EDIT: This is all about the "mpc85xx: Fix panic on SMP init via kernel patch".

Unfortunately, the problem here is that the code and all is coming from a different author (Xiaoming Ni). If you take his patch, please keep the original author and sign-offs in place.
This has to do with upstream policies. (which were set in place to prevent a SCO redux)

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin

(I guess it's going to be quicker if I just push the version I sent you - Thanks for letting us know about the SMP issue though, I don't have a mpc85xx device so I would never got that error.)

Hurricos commented:
> EDIT: This is all about the "mpc85xx: Fix panic on SMP init via kernel patch".
> 
> Unfortunately, the problem here is that the code and all is coming from a different author (Xiaoming Ni). If you take his patch, please keep the original author and sign-offs in place. This has to do with upstream policies. (which were set in place to prevent a SCO redux)
> 
> https://www.kernel.org/doc/html/latest/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin
> 
> (I guess it's going to be quicker if I just push the version I sent you - Thanks for letting us know about the SMP issue though, I don't have a mpc85xx device so I would never got that error.)

My bad! Thank you for pointing out that the attached patch is misattributed. I've force-pushed the fix (I downloaded the author's patch email to make the .patch file, hopefully it applies correctly). I am not used to accepting kernel patches so I literally re-made the patch, committed it to a local repo and then `git format-patch -p1`'d to produce the result :(

I also renamed the patch as a 1xx-series patch as it is literally awaiting upstream.
